R/calcCapacityFactorHist.R

```diff
@@ -1,168 +1,120 @@
+  hist_gen <- readSource(type = "IRENA", subtype = "Generation") # Units are GWh
+
+  # Calculate 2015 capacity factor for relevant technologies, using 2*generation(t) / (cap(t) + cap(t-1) )
+  # then average over 5 years for 2015
+  
+  # load 5 years of generation
+  hist_gen2 = hist_gen[,seq(2013,2017,1), mapping_Irena$irena]%>% 
+    as.data.frame() 
+  
+  # load 5 years of capacity
+  hist_cap2a = hist_cap[,seq(2012,2016,1), mapping_Irena$irena] %>% 
+    as.data.frame() %>% 
+    mutate(Year = as.factor(as.integer(as.character(.data$Year)) + 1)) %>% 
+    select(.data$Year, .data$Region, .data$Data1, 'cap1' ='Value')
+  
+  # load 5 years of capacity
+  hist_cap2b = hist_cap[,seq(2013,2017,1), mapping_Irena$irena]%>% 
+    as.data.frame() %>% 
+    select(.data$Year, .data$Region, .data$Data1, 'cap2' ='Value')
+  
+  hist_cap2 <- full_join(hist_cap2b, hist_cap2a) %>% 
+    mutate(cap = (.data$cap1 + .data$cap2) / 2) %>%
+    select(.data$Year, .data$Region, .data$Data1, .data$cap)
+  
+  cf_year <- full_join(hist_gen2, hist_cap2) %>% 
+    mutate(Value = .data$Value / (8760 * .data$cap))
+    
+  cf_year$Value[cf_year$cap < 0.2] <- 0 # remove those CFs if the installed capacity are too tiny
+  
+  cf_year <- cf_year %>% 
+    select(.data$Year,.data$Region,.data$Data1,.data$Value) %>% 
+    as.magpie()
+  
   #rename
-  getNames(cf_realworld) <- rem_Irena_map$rem
-  #check data
-  max(cf_realworld[,,"hydro"],na.rm = T)
-  max(cf_realworld[,,"wind"],na.rm = T) #INF spm, >1 AZE
-  max(cf_realworld[,,"spv"],na.rm = T)
-  max(cf_realworld[,,"csp"],na.rm = T)
-  max(cf_realworld[,,"bioigcc"],na.rm = T) #>1 CHL, JPN, POL
-  max(cf_realworld[,,"geohdr"],na.rm = T)
-
+  getNames(cf_year) <- mapping_Irena$remind    
   #correct SPM infinite value
-  cf_realworld[is.infinite(cf_realworld)] <- 0.8
+  cf_year[is.infinite(cf_year)] <- 0.8
   #correct AZE,CHL,JPN,POL >1 value
-  cf_realworld[cf_realworld > 1] <- 0.8
+  cf_year[cf_year > 1] <- 0.8
   #get rid of NAs
-  cf_realworld[is.na(cf_realworld)] <- 0
-
-
+  cf_year[is.na(cf_year)] <- 0      
+    
+  # averaging over 5 years for non-0 CFs
+  cf_realworld_n0 <- cf_year %>%
+    as.data.frame() %>% 
+    filter(.data$Value != 0) %>% 
+    group_by(.data$Region, .data$Data1) %>%
+    summarise(Value=mean(.data$Value)) %>%
+    mutate(Year="2015") %>%
+    select(.data$Year,.data$Region,.data$Data1,.data$Value) %>%
+    ungroup()
+  
+  # for regions and techs with 0 CFs for all 5 years
+  cf_realworld_0 <- cf_year %>%
+    as.data.frame() %>% 
+    group_by(.data$Region, .data$Data1) %>%
+    summarise(Value=sum(.data$Value)) %>%
+    filter(.data$Value == 0) %>% 
+    mutate(Year="2015") %>%
+    select(.data$Year,.data$Region,.data$Data1,.data$Value) %>%
+    ungroup()
+  
+  cf_realworld <- full_join(cf_realworld_n0, cf_realworld_0) %>% 
+    group_by(.data$Region, .data$Year) %>%
+    mutate(Data1 = factor(.data$Data1, levels=mapping_Irena$irena) ) %>% 
+    ungroup() %>% 
+    as.magpie()
```

Comment:
this creates duplicates (see warning). so most likely sth is not working as expected. ultimately, this function throws an error, and the cause is probably in these lines

```
~ WARNING: 
~ Duplicate entries found, only the last entry will be used (duplicate entries: ARG|2015|NA, AUS|2015|NA, AUT|2015|NA, BEL|2015|NA, BGR|2015|NA, BRA|2015|NA, CAN|2015|NA, CHE|2015|NA, CHL|2015|NA, CHN|2015|NA, [..]
```